engine/gamesys/src/gamesys/scripts/script_resource.cpp

```diff
@@ -838,6 +874,707 @@ static int SetTexture(lua_State* L)
+ *     }
+ *     table.insert(data.animations, my_animation)
+ *     resource.set_atlas("/main/my_atlas.a.texturesetc", data)
+ * end
+ * ```
+ *
+ * @examples
+ * Sets atlas data for a 256x256 texture with a single animation being rendered as a quad
+ *
+ * ```lua
+ * function init(self)
+ *     local params = {
+ *         texture = "/main/my_256x256_texture.texturec",
+ *         animations = {
+ *             {
+ *                 id          = "my_animation",
+ *                 width       = 256,
+ *                 height      = 256,
+ *                 frame_start = 1,
+ *                 frame_end   = 2,
+ *             }
+ *         },
+ *         geometries = {
+ *             {
+ *                 vertices = {
+ *                     0,   0,
+ *                     0,   256,
+ *                     256, 256,
+ *                     256, 0
+ *                 },
+ *                 uvs = {
+ *                     0, 0,
+ *                     0, 256,
+ *                     256, 256,
+ *                     256, 0
+ *                 },
+ *                 indices = { 0,1,2,0,2,3 }
+ *             }
+ *         }
+ *     }
+ *     resource.set_atlas("/main/test.a.texturesetc", params)
+ * end
+ * ```
+ */
+
+static int SetAtlas(lua_State* L)
+{
+    DM_LUA_STACK_CHECK(L, 0);
+
+    dmhash_t path_hash = dmScript::CheckHashOrString(L, 1);
+    void* atlas_res    = CheckResource(L, g_ResourceModule.m_Factory, path_hash, "texturesetc");
+    assert(atlas_res);
+
+    dmGameSystemDDF::TextureSet texture_set_ddf = {};
+    uint32_t frame_index_count                  = 0;
+    uint32_t num_geometries                     = 0;
+    uint32_t num_animations                     = 0;
+    float tex_width                             = 0;
+    float tex_height                            = 0;
+
+    luaL_checktype(L, 2, LUA_TTABLE);
+    lua_pushvalue(L, 2);
+
+    // Set texture resource path
+    lua_getfield(L, -1, "texture");
+    {
+        const char* texture_path         = luaL_checkstring(L, -1);
+        dmhash_t tex_canonical_path_hash = GetCanonicalPathHash(texture_path);
+        void* texture_res                = CheckResource(L, g_ResourceModule.m_Factory, tex_canonical_path_hash, "texturec");
+        assert(texture_res);
+        texture_set_ddf.m_Texture        = texture_path;
+        dmGraphics::HTexture texture     = (dmGraphics::HTexture) texture_res;
+        tex_width                        = dmGraphics::GetTextureWidth(texture);
+        tex_height                       = dmGraphics::GetTextureHeight(texture);
+    }
+    lua_pop(L, 1); // "texture"
+
+    // Note: We do separate passes over the lua state to validate the data in the args table,
+    //       this is because we need to allocate dynamic memory and can't use luaL_check** functions
+    //       since they longjmp away so we can't release the memory..
+    ValidateSetAtlasGeometries(L, &num_geometries);
+    ValidateSetAtlasAnimations(L, &num_animations);
```

Comment:
In the first iteration I didn't do this pre-pass but instead did the checks manually in-place (using lua_istable and so on). But it was a lot more unreadable so decided to go for this way instead but perhaps there's a better option? The dynamic memory part is annoying, because we will have leaks if errors happen as the comment states..

Reply:
I think we don't generally have a strategy for gracefully supporting memory leaks if a lua error happens.
Personally, I think it's ok, since I also expect the user to do it correctly, and _not_ generate an error.